lib/CL/poclcc_binary.c

```diff
@@ -0,0 +1,443 @@
+#include "poclcc_binary.h"
+
+void poclcc_free(poclcc_global *binary_format){
+  if (binary_format != NULL)
+    if (binary_format->devices != NULL)
+      {
+        int i;
+        for (i=0; i<binary_format->num_devices; i++)
+          poclcc_device_free(&(binary_format->devices[i]));
+        POCL_MEM_FREE(binary_format->devices);
+      }
+}
+
+void poclcc_device_free(poclcc_device *device){
+  if (device != NULL)
+    if (device->kernels != NULL)
+      {
+        int j;
+        for (j=0; j<device->num_kernels; j++)
+          poclcc_kernel_free(&(device->kernels[j]));
+        POCL_MEM_FREE(device->kernels);
+      }
+}
+
+void poclcc_kernel_free(poclcc_kernel *kernel){
+  if (kernel != NULL)
+    {
+      POCL_MEM_FREE(kernel->kernel_name);
+      POCL_MEM_FREE(kernel->binary);
+      POCL_MEM_FREE(kernel->dyn_arguments);
+      POCL_MEM_FREE(kernel->arg_info);
+    }
+}
+
+
+int poclcc_check_global(poclcc_global *binary_format){
+  if (binary_format->version != POCLCC_VERSION)
+    return 0;
+  return !strncmp(binary_format->pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH);
+}
+
+int poclcc_check_device(poclcc_device *device){
+  if (device->version != POCLCC_VERSION)
+    return 0;
+  return !strncmp(device->pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH);
+}
+
+#define FNV_OFFSET UINT64_C(0xcbf29ce484222325)
+#define FNV_PRIME UINT64_C(0x100000001b3)
+uint64_t poclcc_get_device_id(cl_device_id device){
+  //FNV-1A with vendor_id, llvm_target_triplet and llvm_cpu
```

Comment:
Is there a need to hash these? What about just appending them together? It would be more readable in the contents?